File: Pre-Written Code/Data Structures/2D Prefix Sums.cpp
#include <iostream>
#include <vector>
using namespace std;
typedef long long ll;

struct PrefSum2D {
	vector<vector<ll>> pref; int n; PrefSum2D() {}
	PrefSum2D(vector<vector<ll>> &_v) : n(_v.size()), pref(_v) {
		for (int j = 1; j < n; j++) { pref[0][j] += pref[0][j - 1]; }
		for (int i = 1; i < n; i++) { pref[i][0] += pref[i - 1][0]; }
		for (int i = 1; i < n; i++)	for (int j = 1; j < n; j++)
				pref[i][j] += pref[i - 1][j] + pref[i][j - 1] - pref[i - 1][j - 1];
	}
	ll query(int xTL, int yTL, int xBR, int yBR) {
		ll res = pref[xBR][yBR];
		if (xTL) { res -= pref[xTL - 1][yBR]; }
		if (yTL) { res -= pref[xBR][yTL - 1]; }
		if (xTL && yTL) { res += pref[xTL - 1][yTL - 1]; }
		return res;
	}
};

int main() {
	ios::sync_with_stdio(0);
	cin.tie(0); cout.tie(0);

}

File: Pre-Written Code (Python)/Data Structures/2D Prefix Sums.py
class PrefSum2D:
    def __init__(self, v):
        self.n = len(v)
        self.pref = v
        j = 1
        while j < self.n:
            self.pref[0][j] += self.pref[0][j-1]
            j += 1
        i = 1
        while i < self.n:
            self.pref[i][0] += self.pref[i-1][0]
            i += 1
        i = 1
        while i < self.n:
            j = 1
            while j < self.n:
                self.pref[i][j] += self.pref[i-1][j] + self.pref[i][j-1] - self.pref[i-1][j-1]
                j += 1
            i += 1
    def query(self, xTL, yTL, xBR, yBR):
        res = self.pref[xBR][yBR]
        if xTL:
            res -= self.pref[xTL-1][yBR]
        if yTL:
            res -= self.pref[xBR][yTL-1]
        if xTL and yTL:
            res += self.pref[xTL-1][yTL-1]
        return res
        
